### PR TITLE
[Fix](column predicate) seperate CHAR primitive type for column predicate

### PR DESCRIPTION
**Pull request by airborne12**
## Proposed changes

Issue Number: close #xxx

The test inverted_index_p0/test_index_like_select fails when the flags enable_low_cardinality_optimize=false and enable_function_pushdown=true are set. The SQL statement causing the failure is:
```
select * from index_like_select where studentInfo like "%hobbies:" order by name;
```

table index_like_select schema is 

```
CREATE TABLE `index_like_select` (
  `name` varchar(50) NULL,
  `grade` varchar(30) NOT NULL,
  `fatherName` varchar(50) NULL,
  `matherName` varchar(50) NULL,
  `otherinfo` varchar(100) NULL,
  `age` int(11) NOT NULL,
  `registDate` datetime NULL,
  `studentInfo` char(100) NULL,
  `tearchComment` text NULL,
  `selfComment` text NULL,
  INDEX name_idx (`name`) USING INVERTED COMMENT 'name index',
  INDEX grade_idx (`grade`) USING INVERTED PROPERTIES("parser" = "none") COMMENT 'grade index',
  INDEX fatherName_idx (`fatherName`) USING INVERTED PROPERTIES("parser" = "standard") COMMENT ' fatherName index',
  INDEX age_idx (`age`) USING INVERTED COMMENT 'age index',
  INDEX tearchComment_idx (`tearchComment`) USING INVERTED PROPERTIES("parser" = "english") COMMENT 'tearchComment index',
  INDEX studentInfo_idx (`studentInfo`) USING INVERTED PROPERTIES("parser" = "standard") COMMENT 'studentInfo index',
  INDEX selfComment_idx (`selfComment`) USING INVERTED PROPERTIES("parser" = "standard") COMMENT 'selfComment index'
) ENGINE=OLAP
DUPLICATE KEY(`name`)
COMMENT 'OLAP'
DISTRIBUTED BY HASH(`name`) BUCKETS 10
PROPERTIES (
"replication_allocation" = "tag.location.default: 1",
"is_being_synced" = "false",
"storage_format" = "V2",
"light_schema_change" = "true",
"disable_auto_compaction" = "false",
"enable_single_replica_compaction" = "false"
); 

```
Root cause is as following:
When the column type is CHAR, the like predicate processing fails to trim the padding tail value represented by \0.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

File: be/src/vec/olap/block_reader.cpp [hunk 1]
```diff
@@ -503,7 +503,7 @@ bool BlockReader::_get_next_row_same() {
     }
 }
 
-ColumnPredicate* BlockReader::_parse_to_predicate(const FunctionFilter& function_filter) {
+/*ColumnPredicate* BlockReader::_parse_to_predicate(const FunctionFilter& function_filter) {
```
Comment (yiguolei) [inline be/src/vec/olap/block_reader.cpp:506]: 直接把代码删了吧。 否则以后当我们重构代码的时候，会发现这个被注释的代码引用了一些代码，但是实际上没用，会产生误导。

Comment (airborne12) [reply be/src/vec/olap/block_reader.cpp:506]: OK

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.77 seconds
 stream load tsv:          538 seconds loaded 74807831229 Bytes, about 132 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          28.9 seconds inserted 10000000 Rows, about 346K ops/s
 storage size: 17161982804 Bytes

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 48.2 seconds
 stream load tsv:          549 seconds loaded 74807831229 Bytes, about 129 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17161907673 Bytes

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 48.86 seconds
 stream load tsv:          546 seconds loaded 74807831229 Bytes, about 130 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17162000834 Bytes

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.28 seconds
 stream load tsv:          552 seconds loaded 74807831229 Bytes, about 129 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17161945646 Bytes

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 48.44 seconds
 stream load tsv:          552 seconds loaded 74807831229 Bytes, about 129 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17162159373 Bytes

**Review by yiguolei — DISMISSED** [review]
LGTM

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.76 seconds
 stream load tsv:          538 seconds loaded 74807831229 Bytes, about 132 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          28.8 seconds inserted 10000000 Rows, about 347K ops/s
 storage size: 17162036625 Bytes

File: be/src/common/config.cpp [hunk 2]
```diff
@@ -324,7 +324,7 @@ DEFINE_mInt32(data_page_cache_stale_sweep_time_sec, "300");
 DEFINE_mInt32(index_page_cache_stale_sweep_time_sec, "600");
 DEFINE_mInt32(pk_index_page_cache_stale_sweep_time_sec, "600");
 
-DEFINE_Bool(enable_low_cardinality_optimize, "true");
+DEFINE_mBool(enable_low_cardinality_optimize, "true");
```
Comment (yiguolei) [inline be/src/common/config.cpp:327]: 这个不要 modify了，可能有风险，可能会core

**Review by yiguolei — APPROVED** [review]
LGTM

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.7 seconds
 stream load tsv:          537 seconds loaded 74807831229 Bytes, about 132 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17162062411 Bytes

**Review by yiguolei — APPROVED** [review]
LGTM

File: be/src/runtime/primitive_type.h [hunk 3]
```diff
@@ -91,6 +91,10 @@ constexpr bool is_string_type(PrimitiveType type) {
     return type == TYPE_CHAR || type == TYPE_VARCHAR || type == TYPE_STRING;
 }
 
+constexpr bool is_variant_string_type(PrimitiveType type) {
```
Comment (xiaokang) [inline be/src/runtime/primitive_type.h:94]: variant string is ambiguous to variant type. can we use is_var_string_type or is_varchar_type

File: be/src/vec/columns/predicate_column.h [hunk 4]
```diff
@@ -154,8 +153,16 @@ class PredicateColumnType final : public COWHelper<IColumn, PredicateColumnType<
 
     size_t size() const override { return data.size(); }
 
-    [[noreturn]] StringRef get_data_at(size_t n) const override {
-        LOG(FATAL) << "get_data_at not supported in PredicateColumnType";
+    StringRef get_data_at(size_t n) const override {
+        if constexpr (std::is_same_v<T, StringRef>) {
+            auto res = reinterpret_cast<const StringRef&>(data[n]);
+            if constexpr (Type == TYPE_CHAR) {
+                res.size = strnlen(res.data, res.size);
```
Comment (xiaokang) [inline be/src/vec/columns/predicate_column.h:160]: strnlen may be less then CHAR schema length.

Comment (airborne12) [reply be/src/vec/columns/predicate_column.h:160]: it has to be trimed \0

File: be/src/olap/like_column_predicate.cpp [hunk 5]
```diff
@@ -83,18 +91,18 @@ uint16_t LikeColumnPredicate::evaluate(const vectorized::IColumn& column, uint16
...
             }
         } else {
-            auto* str_col =
-                    vectorized::check_and_get_column<vectorized::PredicateColumnType<TYPE_STRING>>(
-                            nested_col);
+            auto* str_col = vectorized::check_and_get_column<vectorized::PredicateColumnType<T>>(
+                    nested_col);
             if (!nullable_col->has_null()) {
                 vectorized::ColumnUInt8::Container res(size, 0);
-                (_state->predicate_like_function)(
```
[the remark this answers is not in the document]
Comment (xiaokang) [reply be/src/olap/like_column_predicate.cpp:91]: get

**Review by xiaokang — APPROVED** [review]
LGTM

**Pull request merged**